### ISSUE TITLE
fix in memory implementation read isolation

### ISSUE THREAD

**Issue by mschoch** [blevesearch/bleve#118]
The KVReader returned by the KVStore.Reader() method is supposed to be isolated from write operations, however the in memory implementation does not honor this contract.

**Comment by sacheendra** [comment]
This test which checks for reader isolation passes for the in memory option. 
https://github.com/blevesearch/bleve/blob/master/index/store/inmem/store_test.go#L109
I do not understand how isolation is implemented in the in memory store. 
I don't see any locking based or snapshot based isolation in both reader.go and store.go
Can you explain how the isolation works?

**Comment by sacheendra** [comment]
Just found that the the function referenced above is not being called at all.

**Comment by mschoch** [comment]
It is not implemented at all, that is the bug.

Historically we did not have reader isolation at all.  Then later we added it.  Unfortunately, the current in-memory only implementation cannot easily be extended to offer reader isolation.  Sure, we could add a lock so that all reads and writes are serialized, but I'd rather we choose a new in-memory implementation that does some copy-on-write tricks to more easily allow reader isolation.

**Comment by mschoch** [comment]
Yes, the problem was that all these tests were once shared, but it turned out to be inconvenient to try and share this code.

When they were copied into the kvstores directly that method was copied even though the in-memory test doesn't call it, because it is known to be broken.

**Comment by sacheendra** [comment]
As there is an in memory implementation(gtreap), can this be closed. A treap is faster than a skip-list, so shouldn't gtreap be made the default inmem implementation.

**Comment by mschoch** [comment]
Yes probably.  There is some active development going on in this area.  @steveyen has been experimenting with different options and benchmarking their performance.

I belive gtreap does pass the reader isolation test, and is about the same performance-wise as the current in-mem implementation.  So, from my perspective it probably should become the default.

However, we've also observed that at the moment, they are both slower than leveldb.  Leveldb has some advanced memory usage, but it still seems unusual and we're investigating.  In this process @steveyen found the cznic in memory btree.  This is faster than leveldb, but doesn't look like it can easily support reader isolation.

In summary, I would say its still an open topic.  But, trying to add in reader isolation to the current skiplist implementation probably isn't going to yield the best solution.

**Comment by sacheendra** [comment]
@steveyen Can you share how you are benchmarking the stores? Can I help in any way?
I don't understand one thing about the gtreap implementation, it looks like multiple writers can make changes to the store at once. If multiple writers are executing batches at the same time, wont there be loss of data? 
My understanding the Execute function (https://github.com/blevesearch/bleve/blob/master/index/store/gtreap/gtreap.go#L291) is as follows:
1. Get a copy of the treap. 
2. Apply changes
3. Store the new treap in store
If another writer was doing operations during the apply changes stage, wont those be lost after we store this treap?
Am I getting something wrong?

**Comment by sacheendra** [comment]
@steveyen Can you share how you are benchmarking the stores? Can I help in any way?
I don't understand one thing about the gtreap implementation, it looks like multiple writers can make changes to the store at once. If multiple writers are executing batches at the same time, wont there be loss of data? 
My understanding the Execute function (https://github.com/blevesearch/bleve/blob/master/index/store/gtreap/gtreap.go#L291) is as follows:
1. Get a copy of the treap. 
2. Apply changes
3. Store the new treap in store
If another writer was doing operations during the apply changes stage, wont those be lost after we store this treap?
Am I getting something wrong?

**Comment by steveyen** [comment]
Hi @sacheendra 

On benchmarking, there are two things...  First off, for microbenchmarks, I'd just followed the benchmarking structure laid out originally by @mschoch in the https://github.com/blevesearch/bleve/tree/master/index/upside_down directory.  There are lots of benchmark_STORE.go files there, which I imagine will need improvement over time.  Right now, for example, I think the microbenchmarks are kind of index-heavy.

Secondly, there's another new project, bleve-bench (again created by @mschoch), that uses wikipedia data to attempt to get more real-world inspired perf characterization: https://github.com/blevesearch/bleve-bench

> I don't understand one thing about the gtreap implementation, it looks like multiple writers can make changes to the store at once. If multiple writers are executing batches at the same time, wont there be loss of data? 

That's true that multiple writers might be operating concurrently, but near the end of the Execute function...

    https://github.com/blevesearch/bleve/blob/master/index/store/gtreap/gtreap.go#L336

...the idea is each writer tries to do a poor-man's compare-&-swap step to see if it "won".  So, it's an optimistic kind of concurrency approach.

**Comment by mschoch** [comment]
Closing, as we now use gtreap for in memory indexes.
